Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx

```diff
@@ -0,0 +1,382 @@
+}
+
+//-----------------------------------------------------------------------------
+void qMRMLTextWidget::onTextNodeModified()
+{
+  Q_D(qMRMLTextWidget);
+  d->TextNodeModified = true;
+  this->updateRequested();
+}
+
+//-----------------------------------------------------------------------------
+void qMRMLTextWidget::updateWidget()
+{
+  Q_D(qMRMLTextWidget);
+
+  bool updateText = true;
+  bool editing = d->isEditing();
+  if (this->ReadOnly)
+    {
+    updateText = true;
+    d->TextEdit->setReadOnly(true);
+    d->EditButton->setVisible(false);
+    d->CancelButton->setVisible(false);
+    d->SaveButton->setVisible(false);
+    }
+  else if (this->AutoUpdate)
+    {
+    updateText = true;
+    d->TextEdit->setReadOnly(false);
+    d->EditButton->setVisible(false);
+    d->CancelButton->setVisible(false);
+    d->SaveButton->setVisible(false);
+    }
+  else if (editing)
+    {
+    updateText = false;
+    d->TextEdit->setReadOnly(false);
+    d->EditButton->setVisible(false);
+    d->CancelButton->setVisible(true);
+    d->SaveButton->setVisible(true);
+    }
+  else
+    {
+    updateText = true;
+    d->TextEdit->setReadOnly(true);
+    d->EditButton->setVisible(true);
+    d->CancelButton->setVisible(false);
+    d->SaveButton->setVisible(false);
+    }
+
+  d->EditButton->setEnabled(!editing);
+  d->CancelButton->setEnabled(editing);
+  d->SaveButton->setEnabled(editing);
+
+  bool wasBlocking = d->TextEdit->blockSignals(true);
+  if (!d->CurrentTextNode)
+    {
+    d->TextEdit->setReadOnly(true);
+    d->TextEdit->setText("");
+    d->TextNodeModified = false;
+    d->EditButton->setEnabled(false);
+    d->CancelButton->setEnabled(false);
+    d->SaveButton->setEnabled(false);
+    }
+  else if (updateText)
+    {
+    int position = d->TextEdit->textCursor().position();
+    std::string text = d->CurrentTextNode->GetText();
+    d->TextEdit->setText(text.c_str());
+    d->TextNodeModified = false;
+    QTextCursor cursor = d->TextEdit->textCursor();
+    position = std::min(position, d->TextEdit->toPlainText().length());
+    cursor.setPosition(position);
+    d->TextEdit->setTextCursor(cursor);
+    }
+  d->TextEdit->blockSignals(wasBlocking);
+
+  this->updateFinished();
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::setMRMLScene(vtkMRMLScene* scene)
+{
+  this->Superclass::setMRMLScene(scene);
+  this->updateRequested();
+}
+
+//------------------------------------------------------------------------------
+bool qMRMLTextWidget::isReadOnly()
+{
+  return this->ReadOnly;
+}
+
+//------------------------------------------------------------------------------
+bool qMRMLTextWidget::isEditing()
+{
+  Q_D(qMRMLTextWidget);
+  return d->isEditing();
+}
+
+//------------------------------------------------------------------------------
+void qMRMLTextWidget::setReadOnly(bool readOnly)
+{
+  Q_D(qMRMLTextWidget);
+  if (this->ReadOnly == readOnly)
```

Comment:
What happens if you set read-only while the widget is being edited? Do you save the results? (probably should)